sklearn/metrics/_classification.py

```diff
@@ -321,17 +321,26 @@ def confusion_matrix(y_true, y_pred, *, labels=None, sample_weight=None,
                          "'all', None}")
 
     n_labels = labels.size
-    label_to_ind = {y: x for x, y in enumerate(labels)}
-    # convert yt, yp into index
-    y_pred = np.array([label_to_ind.get(x, n_labels + 1) for x in y_pred])
-    y_true = np.array([label_to_ind.get(x, n_labels + 1) for x in y_true])
+    # If labels are not consecutive integers starting from zero, then
```

Comment:
Since we are improving `confusion_matrix`, could you use `sklearn.utils.validation.check_sample_weight` in the line above (312-315)

Reply:
Hmm, it looks like `check_sample_weight` might hurt efficiency in the case where sample weight is None. Using that code will force us to use the same dtype in both (1) the case where sample_weight is None and (2) the case where sample_weight is specified. And if we are forced to specify a dtype, then it would have to be float64. At least IMO I'd prefer the ability to have an integral confusion matrix, especially in the case where there the samples are not weighted.

I haven't benchmarked yet, but adding this might work against the purpose of this PR. (In general my opinion is that sklearn does far too many of these checks in a way that drastically reduces performance and does not offer a nice way to disable them).

Its possible that just forcing float64 isn't a terrible thing to do, but I'll need to benchmark it. As modifying this line is not a clear improvement, I would prefer if we could merge this (its been >3 years) and look at this issue in a separate PR.

Reply:
@glemaitre So this change might have merit. I was worried that using float64 would impact how fast it took to build the coo_matrix, but it doesn't seem to: 

```python
    import kwplot
    import seaborn as sns
    kwplot.autoplt()
    sns.set()
    ax = sns.lineplot(data=df, x='n', y='time', hue='label')
    ax.set_yscale('log')
    ax.set_xscale('log')

    from sklearn.utils.validation import _check_sample_weight
    import numpy as np
    results = []
    ns = np.logspace(1, 6, 100).astype(np.int)
    for n in ub.ProgIter(ns, desc='time-tradeoff', verbose=3):
        print('n = {!r}'.format(n))

        n_labels = 100
        y_true = np.random.randint(0, n_labels, n).astype(np.int64)
        y_pred = np.random.randint(0, n_labels, n).astype(np.int64)

        sample_weight = np.ones(y_true.shape[0], dtype=np.int64)

        for timer in ti.reset('use-old-uint8-sample-weight-default'):
            with timer:
                if sample_weight.dtype.kind in {'i', 'u', 'b'}:
                    dtype = np.int64
                else:
                    dtype = np.float64
                cm = coo_matrix((sample_weight, (y_true, y_pred)),
                                shape=(n_labels, n_labels), dtype=dtype,
                                ).toarray()
        results.append({
            'n': n,
            'label': ti.label,
            'time': ti.mean(),
        })

        sample_weight = _check_sample_weight(None, y_true, dtype=np.int64)
        for timer in ti.reset('use-new-float64-sample-weight-default'):
            with timer:
                if sample_weight.dtype.kind in {'i', 'u', 'b'}:
                    dtype = np.int64
                else:
                    dtype = np.float64
                cm = coo_matrix((sample_weight, (y_true, y_pred)),
                                shape=(n_labels, n_labels), dtype=dtype,
                                ).toarray()
        results.append({
            'n': n,
            'label': ti.label,
            'time': ti.mean(),
        })
    import pandas as pd
    df = pd.DataFrame(results)

    import kwplot
    import seaborn as sns
    kwplot.autoplt()
    sns.set()
    ax = sns.lineplot(data=df, x='n', y='time', hue='label')
    ax.set_yscale('log')
    ax.set_xscale('log')
```

![image](https://user-images.githubusercontent.com/3186211/110738304-070e0380-81fd-11eb-8ed6-2d8e53905ece.png)


However, the amount of time it takes to actually call `_check_sample_weight` is much greater than the way the code currently is: 

```
    import ubelt as ub

    from sklearn.utils.validation import _check_sample_weight
    import numpy as np
    results = []
    ns = np.logspace(1, 6, 100).astype(np.int)
    for n in ub.ProgIter(ns, desc='time-tradeoff', verbose=3):
        print('n = {!r}'.format(n))

        y_true = np.random.randint(0, 100, n).astype(np.int64)
        y_pred = np.random.randint(0, 100, n).astype(np.int64)
        sample_weight = np.random.rand(n)

        import timerit
        ti = timerit.Timerit(9, bestof=3, verbose=2)
        for timer in ti.reset('old-sample-weight-given'):
            with timer:
                np.asarray(sample_weight)
        results.append({
            'n': n,
            'label': ti.label,
            'time': ti.mean(),
        })

        for timer in ti.reset('new-sample-weight-given'):
            with timer:
                _check_sample_weight(sample_weight, y_true, dtype=np.int64)
        results.append({
            'n': n,
            'label': ti.label,
            'time': ti.mean(),
        })

        for timer in ti.reset('old-sample-weight-default'):
            with timer:
                np.ones(y_true.shape[0], dtype=np.int64)
        results.append({
            'n': n,
            'label': ti.label,
            'time': ti.mean(),
        })

        for timer in ti.reset('new-sample-weight-default'):
            with timer:
                _check_sample_weight(None, y_true, dtype=np.int64)
        results.append({
            'n': n,
            'label': ti.label,
            'time': ti.mean(),
        })

    import pandas as pd
    df = pd.DataFrame(results)

    import kwplot
    import seaborn as sns
    kwplot.autoplt()
    sns.set()
    ax = sns.lineplot(data=df, x='n', y='time', hue='label')
    ax.set_yscale('log')
    ax.set_xscale('log')
```

![image](https://user-images.githubusercontent.com/3186211/110738399-3755a200-81fd-11eb-9509-c5f079ffcd87.png)

But they do seem to converge towards each other as the number of items grows large. Perhaps it's reasonable to make this replacement. 

I just ran all my benchmarks using `_check_sample_weight` and the speed isn't much different, so it makes sense to make this change for readability and API consistency. I'll patch that in.

Reply:
I made a mistake, it is noticeably different, but its not that significant. The old way averages out at about 0.1% of the function computation and the new way is 3x slower, taking about 0.3%. However, the computation doesn't grow as fast as other parts of the code, so this will never be a bottleneck so I'm still for replacing the 4 lines  with 1 function call. 

It would be nice in the future if `_check_sample_weight` could return a weight array with dtype uint8. While it may not have much speed benefit, it would be more memory efficient.

Reply:
Ok, I was wrong again. It is causing a noticeable slowdown to the point where I'm not ok with it anymore. In my previous test, I ran one variant right after the other, so it never hit all of its cases. 

With the original way we have:

```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    82       432       1031.0      2.4      0.1      if sample_weight is None:
    83       216       3294.0     15.2      0.2          sample_weight = np.ones(y_true.shape[0], dtype=np.int64)
    84                                               else:
    85       216        843.0      3.9      0.0          sample_weight = np.asarray(sample_weight)
```

But with the new way we have: 

```
   201       432      40067.0     92.7      3.9      sample_weight = _check_sample_weight(sample_weight, y_true, dtype=np.uint8)
```

That's going from 0.3% to 3.9%. That's too big of a jump for my taste, I'm going to revert to the previous method.

Reply:
Are you sure this has any meaningful impact in practice? Looking at the code of `_check_sample_weight`, it should be efficient enough. Try without the profiler enabled on a large enough on a call to `confusion_matrix` on a dataset that is large enough for performance to matter (e.g. at least one second).

Reply:
Here some results on my machine:

```
(dev) ogrisel@mba scikit-learn % git diff
diff --git a/sklearn/metrics/_classification.py b/sklearn/metrics/_classification.py
index b4ab145d8..35c5aab5f 100644
--- a/sklearn/metrics/_classification.py
+++ b/sklearn/metrics/_classification.py
@@ -34,6 +34,7 @@ from ..utils import assert_all_finite
 from ..utils import check_array
 from ..utils import check_consistent_length
 from ..utils import column_or_1d
+from ..utils.validation import _check_sample_weight
 from ..utils.multiclass import unique_labels
 from ..utils.multiclass import type_of_target
 from ..utils.validation import _num_samples
@@ -312,12 +313,9 @@ def confusion_matrix(y_true, y_pred, *, labels=None, sample_weight=None,
         elif len(np.intersect1d(y_true, labels)) == 0:
             raise ValueError("At least one label specified must be in y_true")
 
-    if sample_weight is None:
-        sample_weight = np.ones(y_true.shape[0], dtype=np.int64)
-    else:
-        sample_weight = np.asarray(sample_weight)
-
-    check_consistent_length(y_true, y_pred, sample_weight)
+    check_consistent_length(y_true, y_pred)
+    sample_weight = _check_sample_weight(sample_weight, y_true,
+                                         dtype=np.float64)
 
     if normalize not in ['true', 'pred', 'all', None]:
         raise ValueError("normalize must be one of {'true', 'pred', "
(dev) ogrisel@mba scikit-learn % python  ~/tmp/time_confusion_matrix.py
1.399 +/- 0.005 s
(dev) ogrisel@mba scikit-learn % git stash 
Saved working directory and index state WIP on speedup_confusion_matrix: 4b8e1e864 Use faster intersect1d instead of list comprehension
(dev) ogrisel@mba scikit-learn % python  ~/tmp/time_confusion_matrix.py
1.408 +/- 0.005 s
```

So no significant difference. It's in the noise.

Reply:
Actually thinking about it, it's fine to keep the code the way it is.

If no `sample_weight` it's passed we would like to have `int64` to allow for precise integer outputs.

If the user is passing float `sample_weight` we do not want to cast them to integer because that would be wrong (and the output will be floating points but that's fine).

Using `_check_sample_weight` will either break on or the other of the above statements. I therefore think the current code is fine. We could probably add a test to check this behavior but this is out of the scope of this PR.